File: sdks/kotlin/sample/src/main/kotlin/io/sild/sample/DevBackend.kt
package io.sild.sample

import io.sild.core.TokenProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request

// DevBackend stands in for the host app's own backend, exactly like web/demo.html:
// it mints a user token and (for the driver chat) asks the dev server to ensure a
// peer conversation. A real host would mint tokens with its API key server-side
// and create peer conversations via POST /v1/conversations (open_assignment:false).
//
// BASE defaults to the Android emulator's alias for the host machine's localhost
// (10.0.2.2), where `make -C backend run` / sild-dev listens on :8080.
object DevBackend {
    const val BASE = "http://10.0.2.2:8080"
    const val USER_ID = "u_demo_android"

    // The trip whose driver↔rider peer chat the "Message driver" card opens. sild-dev
    // derives the driver's id deterministically as "u_driver_<reference>", so the E2E
    // test can act as the driver to round-trip a message against the rider's UI.
    const val DRIVER_TRIP_REF = "trip_9021"

    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    /** A TokenProvider minting a dev token for [userId] (host backend stand-in). */
    fun tokenProviderFor(userId: String) = TokenProvider {
        withContext(Dispatchers.IO) {
            val req = Request.Builder().url("$BASE/v1/dev/widget-token?user_id=$userId").build()
            http.newCall(req).execute().use {
                json.parseToJsonElement(it.body!!.string()).jsonObject.getValue("token").jsonPrimitive.content
            }
        }
    }

    /** A TokenProvider hitting the dev token-mint endpoint (host stand-in). */
    val tokenProvider = tokenProviderFor(USER_ID)

    /** Ensure the trip's driver↔rider peer conversation exists; return its id. */
    suspend fun ensureDriverConversation(reference: String): String = withContext(Dispatchers.IO) {
        val req = Request.Builder().url("$BASE/v1/dev/peer-conversation?user_id=$USER_ID&reference=$reference").build()
        http.newCall(req).execute().use {
            json.parseToJsonElement(it.body!!.string()).jsonObject.getValue("conversation_id").jsonPrimitive.content
        }
    }
}

File: sdks/swift/Sample/Sources/DevBackend.swift
import Foundation
import Sild
import SildCore

/// Stands in for the host app's own backend, exactly like the Android sample's
/// DevBackend: it mints a user token and (for the driver chat) asks the dev server to
/// ensure a peer conversation. A real host would mint tokens with its API key
/// server-side and create peer conversations via POST /v1/conversations.
enum DevBackend {
    /// The simulator shares the host's network, so localhost is the dev server.
    /// SILD_BASE_URL overrides it (CI publishes it on the simulator's launchd).
    static let base = ProcessInfo.processInfo.environment["SILD_BASE_URL"].flatMap {
        $0.isEmpty ? nil : $0
    } ?? "http://localhost:8080"

    static let userId = "u_demo_ios"

    /// The trip whose driver↔rider peer chat the "Message driver" card opens. sild-dev
    /// derives the driver's id as "u_driver_<reference>", so a test can act as the
    /// driver and round-trip a message against the rider's UI.
    static let driverTripRef = "trip_ios_9021"

    static func tokenProvider(for userId: String) -> TokenProvider {
        ClosureTokenProvider { try await mintToken(userId) }
    }

    static func mintToken(_ userId: String) async throws -> String {
        let url = URL(string: "\(base)/v1/dev/widget-token?user_id=\(userId)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let token = json?["token"] as? String else {
            throw NSError(domain: "DevBackend", code: 1)
        }
        return token
    }

    /// Ensure the trip's driver↔rider peer conversation exists; return its id.
    static func ensureDriverConversation(_ reference: String) async throws -> String {
        let url = URL(string: "\(base)/v1/dev/peer-conversation?user_id=\(userId)&reference=\(reference)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let id = json?["conversation_id"] as? String else {
            throw NSError(domain: "DevBackend", code: 2)
        }
        return id
    }
}
